### PR TITLE
Allow for custom packages

### PR DESCRIPTION
**Pull request by nabuskey**
This PR allows for end users to specify additional packages to be installed. 

Example command: 

```
./idpbuilder create --package-dir testDir --package-dir testDir2
```
The contents of `testDir2`: 

```
1 directory, 2 files
ubuntu@ip-10-192-11-119:~/idpbuilder$ tree testDir2/
testDir2/
└── exampleApp.yaml

0 directories, 1 file

```

This contains a single manifest that defines an ArgoCD Application. This is passed directly to ArgoCD. 

```
ubuntu@ip-10-192-11-119:~/idpbuilder$ head -n 9 testDir2/exampleApp.yaml
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: guestbook
  namespace: argocd
spec:
  project: default
  source:
    repoURL: https://github.com/argoproj/argocd-example-apps.git
```
image

In `testDir`, it contains `app.yaml`. This file is an ArgoCD application except for one field, `spec.source.repoURL`. 

```
ubuntu@ip-10-192-11-119:~/idpbuilder$ tree testDir/
testDir/
├── app.yaml
└── busybox
    └── busybox.yaml

ubuntu@ip-10-192-11-119:~/idpbuilder$ grep repoURL -B 5 testDir/app.yaml
spec:
  destination:
    namespace: my-app
    server: "https://kubernetes.default.svc"
  source:
    repoURL: cnoe://busybox
```
The field above specifies the local path that contains manifests to be synced to Gitea repository. It will create a Gitea repository, fill it with contents from the busybox directory, then crates an ArgoCD application. 

image

image

This will also continuously sync local file contents to the repository as long as the controller manager does not exit. 

You can also add more ArgoCD application files to the specified directories, and the reconciler will pick them up.

## Notes

I am not sure if the API is what we want. Initially I thought to do something like

```
type ExtraPackageConfigSpec struct {
       // Type specifies what kind of package this is. local means local files specified in the Directory field must be synced to a Git Repository
       Type: string `json:"type"`
	// Directory specifies the absolute path to the directory which contains raw manifests to be installed
	Directory string `json:"directory,omitempty"`
	// ArgoApplicationFile specifies the absolute path to the ArgoCD application file
	ArgoApplicationFile string `json:"argoApplicationFile"`
	ArgoCDName string
	ArgoCDNamespace string
}

```
This means Argo apps are defined in CLI and can't add more apps unless you exit the process and run the command again. Not a deal breaker imo but the other way is more user friendly? I don't personally have a preference so any suggestions here is welcome.

### CODE REVIEW

**Comment by nimakaviani** [comment]
> This means Argo apps are defined in CLI and can't add more apps unless you exit the process and run the command again. Not a deal breaker imo but the other way is more user friendly? I don't personally have a preference so any suggestions here is welcome.

I think we should enable both. ideally what would be effective is that the CLI entries for the `package-dir` get converted to a `PackageConfigSpec` and dropped into the kind cluster, to be picked up later on by Argo CD. it is important and helpful for folks to be able to add packages at runtime without having to stop the whole thing

**Comment by greghaynes** [comment]
I would break out custom package to its own top level resource rather than being a field on local build. I'm less concerned with the CLI operation (we can change that later easier) but by having this as its own resource it should be much easier to add/delete packages while idpbuilder is running via CLI or programatically.

**Comment by greghaynes** [comment]
I think we also need to add some docs for how to write these custom packages. It seems like we've got a good handle on the special keyword we're adding to argocd applications at this point.

**Comment by nabuskey** [comment]
> I would break out custom package to its own top level resource rather than being a field on local build. I'm less concerned with the CLI operation (we can change that later easier) but by having this as its own resource it should be much easier to add/delete packages while idpbuilder is running via CLI or programatically.

Yeah I thought about this too. It became a bit messy as I was working through it because of dependency on the local build resource, so I chose the faster way. I will make it a top level resource.

File: pkg/controllers/localbuild/controller.go [hunk 1]
```diff
@@ -385,6 +378,155 @@ func (r *LocalbuildReconciler) shouldShutDown(ctx context.Context, resource *v1a
 	return true, nil
 }
 
+func (r *LocalbuildReconciler) reconcileCustomPkg(ctx context.Context, resource *v1alpha1.Localbuild, pkg v1alpha1.CustomPackageSpec) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+
+	sc := runtime.NewScheme()
```
[the remark this answers is not in the document]
Comment (greghaynes) [reply pkg/controllers/localbuild/controller.go]: sorry I left this a bit ago and forgot to include it in a review - looks like this got fixed though?

**Comment by greghaynes** [comment]
@nabuskey When I run this locally it exits before installing with:

```
1.702074326016658e+09	INFO	setup	Getting Kube config
1.702074326017782e+09	INFO	setup	Getting Kube client
1.702074326037449e+09	INFO	setup	Adding CRDs to the cluster
1.7020743260752127e+09	INFO	crd not yet established, waiting.	{"crd name": "custompackages.idpbuilder.cnoe.io"}
1.7020743260786846e+09	INFO	crd not yet established, waiting.	{"crd name": "custompackages.idpbuilder.cnoe.io"}
1.702074326081224e+09	INFO	crd not yet established, waiting.	{"crd name": "custompackages.idpbuilder.cnoe.io"}
1.7020743261008396e+09	INFO	crd not yet established, waiting.	{"crd name": "gitrepositories.idpbuilder.cnoe.io"}
1.702074326107588e+09	INFO	crd not yet established, waiting.	{"crd name": "gitrepositories.idpbuilder.cnoe.io"}
1.7020743274453123e+09	INFO	setup	Creating controller manager
1.702074327468337e+09	INFO	controller-runtime.metrics	Metrics server is starting to listen	{"addr": ":8080"}
1.70207432746853e+09	INFO	setup	Running controllers
1.7020743274686532e+09	INFO	starting manager
1.7020743274686744e+09	INFO	setup	Creating localbuild resource
1.702074327469005e+09	INFO	Starting server	{"path": "/metrics", "kind": "metrics", "addr": "[::]:8080"}
1.7020743274692569e+09	INFO	controller.localbuild	Starting EventSource	{"reconciler group": "idpbuilder.cnoe.io", "reconciler kind": "Localbuild", "source": "kind source: *v1alpha1.Localbuild"}
1.7020743274692965e+09	INFO	controller.localbuild	Starting Controller	{"reconciler group": "idpbuilder.cnoe.io", "reconciler kind": "Localbuild"}
1.7020743274694698e+09	INFO	controller.custompackage	Starting EventSource	{"reconciler group": "idpbuilder.cnoe.io", "reconciler kind": "CustomPackage", "source": "kind source: *v1alpha1.CustomPackage"}
1.7020743274695013e+09	INFO	controller.custompackage	Starting Controller	{"reconciler group": "idpbuilder.cnoe.io", "reconciler kind": "CustomPackage"}
1.7020743274698658e+09	INFO	controller.gitrepository	Starting EventSource	{"reconciler group": "idpbuilder.cnoe.io", "reconciler kind": "GitRepository", "source": "kind source: *v1alpha1.GitRepository"}
1.7020743274699056e+09	INFO	controller.gitrepository	Starting Controller	{"reconciler group": "idpbuilder.cnoe.io", "reconciler kind": "GitRepository"}
1.702074327669879e+09	INFO	controller.custompackage	Starting workers	{"reconciler group": "idpbuilder.cnoe.io", "reconciler kind": "CustomPackage", "worker count": 1}
1.7020743276709957e+09	INFO	controller.gitrepository	Starting workers	{"reconciler group": "idpbuilder.cnoe.io", "reconciler kind": "GitRepository", "worker count": 1}
1.7020743276709957e+09	INFO	controller.localbuild	Starting workers	{"reconciler group": "idpbuilder.cnoe.io", "reconciler kind": "Localbuild", "worker count": 1}
Error: creating localbuild resource: Localbuild.idpbuilder.cnoe.io "localdev" is invalid: spec.packageConfigs.customPackageDirs: Required value
```

**Comment by nabuskey** [comment]
> @nabuskey When I run this locally it exits before installing with:
> 
> `./idpbuilder create --recreate`
> 
> ```
> 1.702074326016658e+09	INFO	setup	Getting Kube config
> 1.702074326017782e+09	INFO	setup	Getting Kube client
> 1.702074326037449e+09	INFO	setup	Adding CRDs to the cluster
> 1.7020743260752127e+09	INFO	crd not yet established, waiting.	{"crd name": "custompackages.idpbuilder.cnoe.io"}
> 1.7020743260786846e+09	INFO	crd not yet established, waiting.	{"crd name": "custompackages.idpbuilder.cnoe.io"}
> 1.702074326081224e+09	INFO	crd not yet established, waiting.	{"crd name": "custompackages.idpbuilder.cnoe.io"}
> 1.7020743261008396e+09	INFO	crd not yet established, waiting.	{"crd name": "gitrepositories.idpbuilder.cnoe.io"}
> 1.702074326107588e+09	INFO	crd not yet established, waiting.	{"crd name": "gitrepositories.idpbuilder.cnoe.io"}
> 1.7020743274453123e+09	INFO	setup	Creating controller manager
> 1.702074327468337e+09	INFO	controller-runtime.metrics	Metrics server is starting to listen	{"addr": ":8080"}
> 1.70207432746853e+09	INFO	setup	Running controllers
> 1.7020743274686532e+09	INFO	starting manager
> 1.7020743274686744e+09	INFO	setup	Creating localbuild resource
> 1.702074327469005e+09	INFO	Starting server	{"path": "/metrics", "kind": "metrics", "addr": "[::]:8080"}
> 1.7020743274692569e+09	INFO	controller.localbuild	Starting EventSource	{"reconciler group": "idpbuilder.cnoe.io", "reconciler kind": "Localbuild", "source": "kind source: *v1alpha1.Localbuild"}
> 1.7020743274692965e+09	INFO	controller.localbuild	Starting Controller	{"reconciler group": "idpbuilder.cnoe.io", "reconciler kind": "Localbuild"}
> 1.7020743274694698e+09	INFO	controller.custompackage	Starting EventSource	{"reconciler group": "idpbuilder.cnoe.io", "reconciler kind": "CustomPackage", "source": "kind source: *v1alpha1.CustomPackage"}
> 1.7020743274695013e+09	INFO	controller.custompackage	Starting Controller	{"reconciler group": "idpbuilder.cnoe.io", "reconciler kind": "CustomPackage"}
> 1.7020743274698658e+09	INFO	controller.gitrepository	Starting EventSource	{"reconciler group": "idpbuilder.cnoe.io", "reconciler kind": "GitRepository", "source": "kind source: *v1alpha1.GitRepository"}
> 1.7020743274699056e+09	INFO	controller.gitrepository	Starting Controller	{"reconciler group": "idpbuilder.cnoe.io", "reconciler kind": "GitRepository"}
> 1.702074327669879e+09	INFO	controller.custompackage	Starting workers	{"reconciler group": "idpbuilder.cnoe.io", "reconciler kind": "CustomPackage", "worker count": 1}
> 1.7020743276709957e+09	INFO	controller.gitrepository	Starting workers	{"reconciler group": "idpbuilder.cnoe.io", "reconciler kind": "GitRepository", "worker count": 1}
> 1.7020743276709957e+09	INFO	controller.localbuild	Starting workers	{"reconciler group": "idpbuilder.cnoe.io", "reconciler kind": "Localbuild", "worker count": 1}
> Error: creating localbuild resource: Localbuild.idpbuilder.cnoe.io "localdev" is invalid: spec.packageConfigs.customPackageDirs: Required value
> ```
> 
> Are we now making packageDirs a required parameter? What do we think the default command to recommend to users should be?

That was my bad. I did not mean to make that field a required command flag. Fixed now.

**Review by nimakaviani — APPROVED** [review]
some minor changes, otherwise looks good

File: api/v1alpha1/custom_package_types.go [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+type CustomPackageList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []CustomPackage `json:"items"`
+}
+
+// CustomPackageSpec controls the installation of the custom applications.
+type CustomPackageSpec struct {
+	// Replicate specifies whether to replicate remote or local contents to the local gitea server.
+	Replicate bool `json:"replicate"`
```
[the remark this answers is not in the document]
Comment (nabuskey) [reply api/v1alpha1/custom_package_types.go:25]: Yeah it defaults to false via zero value. I'll add an annotation to be explicit.

File: api/v1alpha1/custom_package_types.go [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+type ArgoCDPackageSpec struct {
+	// ApplicationFile specifies the absolute path to the ArgoCD application file
+	ApplicationFile string `json:"applicationFile"`
+	Name            string `json:"name"`
+	Namespace       string `json:"namespace"`
+}
+
+type CustomPackageStatus struct {
+	Synced bool `json:"synced,omitempty"`
```
[the remark this answers is not in the document]
Comment (nabuskey) [reply api/v1alpha1/custom_package_types.go:45]: Good point.

**Comment by cmoulliard** [comment]
Next time, please update the README.md file too ! @nabuskey

**Comment by nimakaviani** [comment]
There is an open issue for it with folks already showing interest https://github.com/cnoe-io/idpbuilder/issues/103 

Manabu chose to let the community help there, if possible. We will pick it up if no traction from the community occurs soon-ish

**Pull request merged**